flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/IntermediateResultPartition.java

```diff
@@ -86,6 +86,11 @@ public boolean isConsumable() {
 	}
 
 	void resetForNewExecution() {
+		if (getResultType().isBlocking() && hasDataProduced) {
+			// A BLOCKING result partition with data produced means it is finished
+			// Need to add the running producer count of the result on resetting it
+			totalResult.incrementNumberOfRunningProducersAndGetRemaining();
+		}
```

Comment:
This logic is a bit unfortunate but atm I guess there is no other way around without having to refactor the whole `IntermediateResultPartition` design. At some point in the future, we should do this, though.

Reply:
Agree